File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Anchor.swift
import CoreFoundation

public struct Anchor: Equatable, Codable {
    public var x: CGFloat
    public var y: CGFloat

    public init(x: CGFloat, y: CGFloat) {
        self.x = x
        self.y = y
    }

    public func translated(dx: CGFloat, dy: CGFloat) -> Anchor {
        Anchor(x: x + dx, y: y + dy)
    }

    public mutating func translate(_ dx: CGFloat, _ dy: CGFloat) {
        x += dx
        y += dy
    }

    public func diff(_ other: Anchor) -> (CGFloat, CGFloat) {
        (x - other.x, y - other.y)
    }

    public func isAt(_ x: CGFloat, _ y: CGFloat) -> Bool {
        abs(self.x - x) <= .ulpOfOne && abs(self.y - y) <= .ulpOfOne
    }

    public func asVector() -> Vector {
        Vector(x: x, y: y)
    }

    public func asPair() -> (CGFloat, CGFloat) {
        (x, y)
    }
}

File: mobile/android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Anchor.kt
package net.olamaelcu.livtet.jigsaw.model

import kotlin.math.abs
import kotlin.random.Random

class Anchor(var x: Float, var y: Float) {

    fun equal(other: Anchor): Boolean = isAt(other.x, other.y)

    fun isAt(x: Float, y: Float): Boolean = this.x == x && this.y == y

    fun translated(dx: Float, dy: Float): Anchor = copy().also { it.translate(dx, dy) }

    fun translate(dx: Float, dy: Float): Anchor {
        x += dx
        y += dy
        return this
    }

    fun closeTo(other: Anchor, tolerance: Float): Boolean =
        abs(x - other.x) <= tolerance && abs(y - other.y) <= tolerance

    fun copy(): Anchor = Anchor(x, y)

    fun diff(other: Anchor): Pair<Float, Float> =
        (x - other.x) to (y - other.y)

    fun asVector(): Vector = Vector(x, y)

    fun export(): Vector = asVector()

    companion object {
        fun anchor(x: Float, y: Float): Anchor = Anchor(x, y)

        fun atRandom(maxX: Float, maxY: Float): Anchor =
            Anchor(Random.nextFloat() * maxX, Random.nextFloat() * maxY)

        fun import(vector: Vector): Anchor = anchor(vector.x, vector.y)
    }
}
